protocol/switch.go

```diff
@@ -0,0 +1,81 @@
+// Package protocol provides core interfaces for protocol routing and negotiation in libp2p.
+package protocol
+
+import (
+	"io"
+)
+
+// HandlerFunc is a user-provided function used by the Router to
+// handle a protocol/stream.
+//
+// Will be invoked with the protocol ID string as the first argument,
+// which may differ from the ID used for registration if the handler
+// was registered using a match function.
+type HandlerFunc = func(protocol string, rwc io.ReadWriteCloser) error
+
+// Router is an interface that allows users to add and remove protocol handlers,
+// which will be invoked when incoming stream requests for registered protocols
+// are accepted.
+//
+// Upon receiving an incoming stream request, the Router will check all registered
+// protocol handlers to determine which (if any) is capable of handling the stream.
+// The handlers are checked in order of registration; if multiple handlers are
+// eligible, only the first to be registered will be invoked.
+type Router interface {
+
+	// AddHandler registers the given handler to be invoked for
+	// an exact literal match of the given protocol ID string.
+	AddHandler(protocol string, handler HandlerFunc)
```

Comment:
why "Add" and not "Set"?
Add says to me that single protocol can be processed by multiple handlers
Set says to me that single protocol can be processed with exactly one handler

Reply:
Yeah this is bad naming, indeed. Could you file a separate issue in this repo so we can consider a rename?